clients/java/src/main/java/io/camunda/zeebe/client/impl/search/ProcessInstanceFilterImpl.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.camunda.zeebe.client.impl.search;
+
+import static io.camunda.zeebe.client.api.search.SearchRequestBuilders.variableValueFilter;
+import static io.camunda.zeebe.client.impl.util.CollectionUtil.addValuesToList;
+import static io.camunda.zeebe.client.impl.util.CollectionUtil.collectValues;
+
+import io.camunda.zeebe.client.api.search.ProcessInstanceFilter;
+import io.camunda.zeebe.client.api.search.VariableValueFilter;
+import io.camunda.zeebe.client.protocol.rest.ProcessInstanceFilterRequest;
+import java.util.List;
+import java.util.function.Consumer;
+
+public class ProcessInstanceFilterImpl
+    extends TypedSearchRequestPropertyProvider<ProcessInstanceFilterRequest>
+    implements ProcessInstanceFilter {
+
+  private final ProcessInstanceFilterRequest filter;
+
+  public ProcessInstanceFilterImpl() {
+    filter = new ProcessInstanceFilterRequest();
+  }
+
+  @Override
+  public ProcessInstanceFilter processInstanceKeys(final Long value, final Long... values) {
+    return processInstanceKeys(collectValues(value, values));
+  }
+
+  @Override
+  public ProcessInstanceFilter processInstanceKeys(final List<Long> values) {
+    filter.setKey(addValuesToList(filter.getKey(), values));
+    return this;
+  }
+
+  @Override
+  public ProcessInstanceFilter variable(final VariableValueFilter value) {
+    final VariableValueFilterImpl variableFilter = (VariableValueFilterImpl) value;
```

Comment:
❓ Do you see a way to avoid those class cast instructions? We have many of them in this PR and I was wondering if we could avoid them. Code style-wise, this is not ideal. There could be a base interface that contains such a method that all the filter, sort, and page interfaces extend.

Reply:
I tried a couple of variations and tried not to expose too much to the public API. Basically, I took the same approach of having a base interface that all classes implement and that provides that necessary method. However, this method would then be available to users as well, which I wanted to avoid. If you think this is not a problem, then okay, I can change that.

Reply:
Yeah, I was also wondering about whether we should expose this or not. I don't have a good answer right now from the top of my head. Exposing it publicly may be even worse 🤔

Reply:
I think, I have an alternative to centralize that. I will finish it quickly; maybe it will make it a bit more handy.

Reply:
Hm... what do you think? 9af53acba28f5dce1f4640f94ea387615e5c16ad At least, it makes it not worse than before.

Reply:
That's a lot better really since it actually checks the base type where this is implemented or extended from 👍